File: platform-gettext/src/test/java/locolaser/platform/mobile/resource/GetTextResourcesTest.java
package locolaser.platform.mobile.resource;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;
import ru.pocketbyte.locolaser.platform.mobile.resource.GetTextResources;
import ru.pocketbyte.locolaser.resource.entity.*;

import java.io.IOException;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

public class GetTextResourcesTest {

    @Rule
    public TemporaryFolder tempFolder= new TemporaryFolder();

    @Test
    public void testWriteAndRead() throws IOException {
        ResMap resMap1 = prepareResMap();

        GetTextResources resources = new GetTextResources(tempFolder.newFolder(), "test");
        resources.write(resMap1, null);

        ResMap resMap2 = resources.read(resMap1.keySet());

        assertNotNull(resMap2);
        assertEquals(resMap1, resMap2);
    }

    //FIXME Map without plurals. Add plural items when platform start support it
    private ResMap prepareResMap() {
        ResMap resMap = new ResMap();
        resMap.put("locale1", prepareResLocale1());
        resMap.put("locale2", prepareResLocale2());
        resMap.put("locale3", prepareResLocale3());
        return resMap;
    }

    private ResLocale prepareResLocale1() {
        ResLocale resLocale = new ResLocale();
        resLocale.put(prepareResItem("key1", new ResValue[]{
                new ResValue("value1_1", null, Quantity.OTHER),
        }));
        resLocale.put(prepareResItem("key2", new ResValue[]{
                new ResValue("value2_1", null, Quantity.OTHER)
        }));
        resLocale.put(prepareResItem("key3", new ResValue[]{
                new ResValue("value3_1", null, Quantity.OTHER),
        }));
        return resLocale;
    }

    private ResLocale prepareResLocale2() {
        ResLocale resLocale = new ResLocale();
        resLocale.put(prepareResItem("key1", new ResValue[]{
                new ResValue("value1_2", null, Quantity.OTHER),
        }));
        resLocale.put(prepareResItem("key3", new ResValue[]{
                new ResValue("value3_2", null, Quantity.OTHER)
        }));
        resLocale.put(prepareResItem("key4", new ResValue[]{
                new ResValue("value4_2", null, Quantity.OTHER),
        }));
        return resLocale;
    }

    private ResLocale prepareResLocale3() {
        ResLocale resLocale = new ResLocale();
        resLocale.put(prepareResItem("key1", new ResValue[]{
                new ResValue("value1_3", null, Quantity.OTHER),
        }));
        resLocale.put(prepareResItem("key4", new ResValue[]{
                new ResValue("value4_3", null, Quantity.OTHER),
        }));
        return resLocale;
    }

    private ResItem prepareResItem(String key, ResValue[] values) {
        ResItem resItem = new ResItem(key);
        for (ResValue value: values)
            resItem.addValue(value);
        return resItem;
    }
}

File: resource-gettext/src/test/java/ru/pocketbyte/locolaser/gettext/resource/GetTextResourcesTest.kt
package ru.pocketbyte.locolaser.gettext.resource

import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder
import ru.pocketbyte.locolaser.resource.entity.*

import java.io.IOException

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import ru.pocketbyte.locolaser.config.ExtraParams

class GetTextResourcesTest {

    @Rule @JvmField
    var tempFolder = TemporaryFolder()

    @Test
    @Throws(IOException::class)
    fun testWriteAndRead() {
        val resMap1 = prepareResMap()

        val resources = GetTextResources(tempFolder.newFolder(), "test", null)
        resources.write(resMap1, null)

        val resMap2 = resources.read(resMap1.keys, ExtraParams())

        assertNotNull(resMap2)
        assertEquals(resMap1, resMap2)
    }

    //FIXME Map without plurals. Add plural items when platform start support it
    private fun prepareResMap(): ResMap {
        val resMap = ResMap()
        resMap["locale1"] = prepareResLocale1()
        resMap["locale2"] = prepareResLocale2()
        resMap["locale3"] = prepareResLocale3()
        return resMap
    }

    private fun prepareResLocale1(): ResLocale {
        val resLocale = ResLocale()
        resLocale.put(prepareResItem("key1", arrayOf(ResValue("value1_1", null, Quantity.OTHER))))
        resLocale.put(prepareResItem("key2", arrayOf(ResValue("value2_1", null, Quantity.OTHER))))
        resLocale.put(prepareResItem("key3", arrayOf(ResValue("value3_1", null, Quantity.OTHER))))
        return resLocale
    }

    private fun prepareResLocale2(): ResLocale {
        val resLocale = ResLocale()
        resLocale.put(prepareResItem("key1", arrayOf(ResValue("value1_2", null, Quantity.OTHER))))
        resLocale.put(prepareResItem("key3", arrayOf(ResValue("value3_2", null, Quantity.OTHER))))
        resLocale.put(prepareResItem("key4", arrayOf(ResValue("value4_2", null, Quantity.OTHER))))
        return resLocale
    }

    private fun prepareResLocale3(): ResLocale {
        val resLocale = ResLocale()
        resLocale.put(prepareResItem("key1", arrayOf(ResValue("value1_3", null, Quantity.OTHER))))
        resLocale.put(prepareResItem("key4", arrayOf(ResValue("value4_3", null, Quantity.OTHER))))
        return resLocale
    }

    private fun prepareResItem(key: String, values: Array<ResValue>): ResItem {
        val resItem = ResItem(key)
        for (value in values)
            resItem.addValue(value)
        return resItem
    }
}
